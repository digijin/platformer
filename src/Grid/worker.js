self.addEventListener(
	"message",
	function(e) {
		var data = e.data;
		switch (data.action) {
			case "repeat":
				self.postMessage("did you say " + data);
				break;
			default:
				self.postMessage({ error: "action required" });
		}
	},
	false
);
