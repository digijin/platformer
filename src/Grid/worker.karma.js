import { setTimeout } from "core-js/library/web/timers";

fdescribe("webworker", () => {
	it("should be defined", () => {
		expect(window.Worker).toBeDefined();
	});
	let worker;
	it("should load worker file and get a message", done => {
		worker = new Worker("/base/src/Grid/worker.js");
		worker.onmessage = e => {
			console.log("received message from worker", e.data);
			done();
		};
		console.log("sending message to worker");
		worker.postMessage({ action: "repeat", data: "abc123" });
	});
	// it("should post a message", () => {});
});
