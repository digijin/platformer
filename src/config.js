//@flow
export default {
	game: {
		width: 800,
		height: 800
	},
	grid: {
		width: 32,
		// height: 20,
		tile: {
			width: 8,
			height: 8
		}
	},
	player: {
		size: {
			w: 40,
			h: 50
		},
		legspeed: 20,
		speed: 500,
		jump: {
			power: 10
		}
	},
	gravity: 20,
	debug: {
		fpsmeter: true,
		player: {
			boundingBox: false
		}
	},
	missile: {
		guidedDist: 50
	},
	fpsmeter: {
		graph: 1,
		theme: "transparent",
		heat: 0,
		history: 20,
		left: "auto",
		right: "5px",
		show: "ms"
	},
	enemy: {
		walkSpeed: 50,
		size: {
			w: 50,
			h: 50
		},
		registration: {
			x: 0.5,
			y: 1
		}
	},
	input: {
		keyboardMapping: {
			a: 65,
			d: 68,
			w: 87,
			s: 83,
			left: 37,
			up: 38,
			right: 39,
			down: 40,
			shift: 16,
			enter: 13,
			ctrl: 17,
			escape: 27,
			space: 32
		},
		axes: {
			horizontal: [
				{
					type: "keyboard",
					positive: "d",
					negative: "a"
				},
				{
					type: "keyboard",
					positive: "right",
					negative: "left"
				},
				{
					type: "gamepad",
					axis: 0
				}
			],
			vertical: [
				{
					type: "keyboard",
					positive: "s",
					negative: "w"
				},
				{
					type: "gamepad",
					axis: 1
				}
			],
			wheel: [
				{
					type: "mouse"
				}
			]
		},
		buttons: {
			boost: [
				{
					type: "keyboard",
					key: "shift"
				}
			],
			down: [
				{
					type: "keyboard",
					key: "down"
				},
				{
					type: "keyboard",
					key: "s"
				}
			],
			jump: [
				{
					type: "gamepad",
					button: 0
				},
				{
					type: "keyboard",
					key: "space"
				}
			],
			special: [
				{
					type: "gamepad",
					button: 1
				},
				{
					type: "gamepad",
					button: "lt"
				},
				{
					type: "mouse",
					button: 2
				}
			],
			fire: [
				{
					type: "gamepad",
					button: 2
				},
				{
					type: "gamepad",
					button: "rt"
				},
				{
					type: "keyboard",
					key: "ctrl"
				},
				{
					type: "mouse",
					button: 0
				}
			],
			editor_add: [
				{
					type: "mouse",
					button: 0
				}
			],
			editor_remove: [
				{
					type: "mouse",
					button: 2
				}
			],
			editor_speed: [
				{
					type: "keyboard",
					key: "shift"
				}
			],
			editor_modifier: [
				{
					type: "keyboard",
					key: "ctrl"
				}
			],
			editor_drag: [
				{
					type: "mouse",
					button: 1
				}
			],
			pause: [
				{
					type: "keyboard",
					key: "escape"
				}
			],
			stand: [
				{
					type: "gamepad",
					button: 5
				}
			]
		}
	}
};
