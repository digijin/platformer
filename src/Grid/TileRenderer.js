import type Engine from "Engine";
import config from "config";

import dirtTile from "dirt_tile.png";

export default class TileRenderer {
	engine: Engine;
	constructor(params: { engine: Engine }) {
		this.engine = params.engine;
	}
	render() {
		this.engine.ctx.context.fillStyle = "#000000";

		//screenRect
		let screenRect = this.engine.ctx.screenRect();
		let blocks = this.engine.grid.getBlocksInRect(screenRect);
		// engine.ctx.strokeStyle = '#000000'd
		blocks.forEach((cell, y) => {
			if (cell.type == "0") {
			} else {
				this.engine.ctx.drawSprite(
					dirtTile,
					cell.point,
					{ w: config.grid.width, h: config.grid.width },
					0,
					{ x: 0, y: 0 }
				);
			}
		});
	}
}
