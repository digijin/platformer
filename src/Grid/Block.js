//@flow

import config from "config";
import Point from "Utility/Point";
import Rect from "Utility/Rect";
import type Grid from "Grid";

import type BlockType from "Grid/Block/Type";
import {
  BlockTypeMap
} from "Grid/Block/Type";

export default class Block {
  position: Point;
  type: string;
  backgroundType: string | void;
  grid: Grid;
  tint: number;
  hp: number;
  constructor(params: {
    position: Point,
    type: string,
    grid: Grid,
    tint ? : number,
    backgroundType ? : string
  }) {
    this.position = params.position;
    this.type = params.type;
    this.grid = params.grid;
    if (params.tint) {
      this.tint = params.tint;
    }
    if (typeof this.tint !== "number") {
      this.tint = 0xffffff;
    }

    if (!params.backgroundType) {
      params.backgroundType = this.type;
    }
    this.backgroundType = params.backgroundType;


    if (this.type !== "0") {
      this.hp = this.getType().hp;
    }
  }
  toString() {
    return JSON.stringify({
      position: this.position,
      type: this.type
    });
  }

  getType(): BlockType {
    return BlockTypeMap[this.type];
  }

  getBackgroundType(): BlockType {
    return BlockTypeMap[this.backgroundType];
  }

  isEmpty(): boolean {
    return this.type == "0";
  }

  isBackgroundEmpty(): boolean {
    return this.backgroundType == "0";
  }
  damage(amount: number) {
    let type = this.getType();
    if (type && type.destructable) {
      this.hp -= amount;
      if (this.hp <= 0) {
        this.destroy();
      }
    }
  }
  destroy() {
    this.type = "0";
    this.grid.bustCache(this);
  }
  //for editor
  add(blockId: string) {
    this.type = blockId;
    this.grid.bustCache(this);
  }
  addBackground(blockId: string) {
    this.backgroundType = blockId;
    this.grid.bustCache(this);
  }
  //for editor
  remove() {
    this.type = "0";
    this.grid.bustCache(this);
  }

  get center(): Point {
    return new Point({
      x: (this.position.x + 0.5) * config.grid.width,
      y: (this.position.y + 0.5) * config.grid.width
    });
  }
  get point(): Point {
    return new Point({
      x: this.position.x * config.grid.width,
      y: this.position.y * config.grid.width
    });
  }
  is(block: Block): boolean {
    return (
      block.position.x === this.position.x &&
      block.position.y === this.position.y
    );
  }

  get rect(): Rect {
    return new Rect({
      t: this.position.y * config.grid.width,
      r: (this.position.x + 1) * config.grid.width,
      b: (this.position.y + 1) * config.grid.width,
      l: this.position.x * config.grid.width
    });
  }

  // get key():string{
  //   return makeKey(this.x, this.y);
  // }

  // static fromPoint(point: Point) {
  // 	return new Block({
  // 		x: Math.floor(point.x / config.grid.width),
  // 		y: Math.floor(point.y / config.grid.width)
  // 	});
  // }
}