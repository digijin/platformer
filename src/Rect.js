//@flow

import { extend } from "lodash";
import config from "config";

import Point from "Point";

import Block from "Block";

export type RenderParams = {
	x: number,
	y: number,
	w: number,
	h: number
};

export default class Rect {
	t: number;
	r: number;
	b: number;
	l: number;
	// static screenRect(){
	// 	let tl = Point.fromScreen(0,0);
	// 	let br = Point.fromScreen(window.innerWidth,window.innerHeight);
	// 	return new Rect(tl, br);
	// }
	static fromPosSizeRego(
		pos: Point,
		size: { w: number, h: number },
		rego: { x: number, y: number }
	) {
		let t = pos.y - size.h * rego.y;
		let l = pos.x - size.w * rego.x;
		return new Rect({
			t: t,
			l: l,
			r: l + size.w,
			b: t + size.h
		});
	}
	overlaps(rect: Rect): boolean {
		let outsideH = this.b <= rect.t || rect.b <= this.t;
		let outsideV = this.r <= rect.l || rect.r <= this.l;
		return !outsideV && !outsideH;
	}
	contains(point: Point): boolean {
		if (point.x < this.l) return false;
		if (point.x > this.r) return false;
		if (point.y < this.t) return false;
		if (point.y > this.b) return false;
		return true;
	}
	width() {
		return this.r - this.l;
	}
	height() {
		return this.b - this.t;
	}
	centerPoint() {
		return new Point({
			x: this.l + this.width() / 2,
			y: this.t + this.height() / 2
		});
	}
	/**
	 * doesnt mutate. returns new moved rect
	 */
	move(amount: { x: number, y: number }) {
		return new Rect({
			t: this.t + amount.y,
			r: this.r + amount.x,
			b: this.b + amount.y,
			l: this.l + amount.x
		});
	}
	/**
	 * Recommended usage to use {t,r,b,l} but can 
	 * be spread or inferred from points
	 * @param {*} params 
	 */
	constructor(params: { t: number, r: number, b: number, l: number }) {
		if (arguments.length === 4) {
			this.t = arguments[0];
			this.r = arguments[1];
			this.b = arguments[2];
			this.l = arguments[3];
		} else if (arguments.length === 1) {
			let a = arguments[0];
			if (a.t !== undefined) {
				extend(this, a);
			} else if (a.x !== undefined) {
				this.t = a.y;
				this.l = a.x;
				this.r = a.x + a.w;
				this.b = a.y + a.h;
			} else {
				throw new Error(
					"Rect constructor given garbage. " + arguments.toString()
				);
			}
		} else if (arguments.length === 2) {
			let p1 = arguments[0];
			let p2 = arguments[1];
			if (p1.x < p2.x) {
				this.l = p1.x;
				this.r = p2.x;
			} else {
				this.l = p2.x;
				this.r = p1.x;
			}
			if (p1.y < p2.y) {
				this.t = p1.y;
				this.b = p2.y;
			} else {
				this.t = p2.y;
				this.b = p1.y;
			}
		}
	}
	blockRect(): Rect {
		return new Rect({
			t: Math.floor(this.t / config.grid.height),
			r: Math.ceil(this.r / config.grid.width),
			b: Math.ceil(this.b / config.grid.width),
			l: Math.floor(this.l / config.grid.height)
		});
	}
	width(): number {
		return this.r - this.l;
	}
	//corners
	tl(): Point {
		return new Point({ x: this.l, y: this.t });
	}
	tr(): Point {
		return new Point({ x: this.r, y: this.t });
	}
	bl(): Point {
		return new Point({ x: this.l, y: this.b });
	}
	br(): Point {
		return new Point({ x: this.r, y: this.b });
	}

	add(rect: { t: number, r: number, b: number, l: number }): Rect {
		return new Rect({
			t: this.t + rect.t,
			r: this.r + rect.r,
			b: this.b + rect.b,
			l: this.l + rect.l
		});
	}

	// get units(): Array<Block> {
	// 	let list = [];
	// 	let sel = this;
	// 	for (let y = sel.t; y <= sel.b; y++) {
	// 		for (let x = sel.l; x <= sel.r; x++) {
	// 			list.push(new Block({ x, y }));
	// 		}
	// 	}
	// 	return list;
	// }
	// get blocks(): Array<Block> {
	// 	let list = [];
	// 	let sel = this.blockRect();
	// 	for (let y = sel.t; y <= sel.b; y++) {
	// 		for (let x = sel.l; x <= sel.r; x++) {
	// 			list.push(new Block({ x, y }));
	// 		}
	// 	}
	// 	return list;
	// }

	// get renderParams():RenderParams {
	// 	let tl = new Point({x:this.l, y:this.t}).screen;
	// 	let br = new Point({x:this.r, y:this.b}).screen;
	// 	return {x: tl.x, y: tl.y, w: br.x-tl.x, h: br.y-tl.y};
	// }
}
