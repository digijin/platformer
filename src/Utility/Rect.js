//@flow

import { extend } from "lodash";
import config from "config";

import Point from "./Point";

import Block from "Grid/Block";

export type RenderParams = {
	x: number,
	y: number,
	w: number,
	h: number
};

export default class Rect {
	t: number;
	r: number;
	b: number;
	l: number;

	static fromPosSizeRego(
		pos: Point,
		size: { w: number, h: number },
		rego: { x: number, y: number }
	) {
		let t = pos.y - size.h * rego.y;
		let l = pos.x - size.w * rego.x;
		return new Rect({
			t: t,
			l: l,
			r: l + size.w,
			b: t + size.h
		});
	}
	static fromPoints(pt1, pt2) {
		return new Rect({
			l: Math.min(pt1.x, pt2.x),
			r: Math.max(pt1.x, pt2.x),
			t: Math.min(pt1.y, pt2.y),
			b: Math.max(pt1.y, pt2.y)
		});
	}
	overlaps(rect: Rect): boolean {
		let outsideH = this.b <= rect.t || rect.b <= this.t;
		let outsideV = this.r <= rect.l || rect.r <= this.l;
		return !outsideV && !outsideH;
	}
	contains(point: Point): boolean {
		if (point.x < this.l) return false;
		if (point.x > this.r) return false;
		if (point.y < this.t) return false;
		if (point.y > this.b) return false;
		return true;
	}
	width() {
		return this.r - this.l;
	}
	height() {
		return this.b - this.t;
	}
	centerPoint() {
		return new Point({
			x: this.l + this.width() / 2,
			y: this.t + this.height() / 2
		});
	}
	/**
	 * doesnt mutate. returns new moved rect
	 */
	move(amount: { x: number, y: number }) {
		return new Rect({
			t: this.t + amount.y,
			r: this.r + amount.x,
			b: this.b + amount.y,
			l: this.l + amount.x
		});
	}
	/**
	 * Recommended usage to use {t,r,b,l} but can
	 * be spread or inferred from points
	 * @param {*} params
	 */
	constructor(params: { t: number, r: number, b: number, l: number }) {
		if (arguments.length === 1) {
			let a = arguments[0];
			if (a.t !== undefined) {
				extend(this, a);
			} else if (a.x !== undefined) {
				this.t = a.y;
				this.l = a.x;
				this.r = a.x + a.w;
				this.b = a.y + a.h;
			} else {
				throw new Error(
					"Rect constructor given garbage. " + arguments.toString()
				);
			}
		} else {
			throw new Error(
				"Rect only takes a single argument, multiple args passed"
			);
		}
	}
	blockRect(): Rect {
		return new Rect({
			t: Math.floor(this.t / config.grid.width),
			r: Math.ceil(this.r / config.grid.width),
			b: Math.ceil(this.b / config.grid.width),
			l: Math.floor(this.l / config.grid.width)
		});
	}
	width(): number {
		return this.r - this.l;
	}
	//corners
	tl(): Point {
		return new Point({ x: this.l, y: this.t });
	}
	tr(): Point {
		return new Point({ x: this.r, y: this.t });
	}
	bl(): Point {
		return new Point({ x: this.l, y: this.b });
	}
	br(): Point {
		return new Point({ x: this.r, y: this.b });
	}

	add(rect: { t: number, r: number, b: number, l: number }): Rect {
		return new Rect({
			t: this.t + rect.t,
			r: this.r + rect.r,
			b: this.b + rect.b,
			l: this.l + rect.l
		});
	}
}
