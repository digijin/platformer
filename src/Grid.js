//@flow

import { EnemyTypesMap } from "Actor/Enemy/Type";
import { Noise } from "noisejs";
import Block from "Grid/Block";
import config from "config";
import Enemy from "Actor/Enemy";
import GameObject from "GameObject";
import Point from "Utility/Point";
import Rect from "Utility/Rect";
import Line from "Utility/Line";
import type Engine from "Engine";

import Decor from "Grid/Decor";
import type DecorType from "Grid/Decor/Type";

export default class Grid extends GameObject {
	blocks: Array<Array<Block>>;

	decor: Array<Decor>;

	z: number;

	highlightBlock(block: Block) {
		if (block) {
			let rect = block.rect;
			this.engine.ctx.context.strokeStyle = "#888";
			this.engine.ctx.strokeRect(
				rect.l,
				rect.t,
				rect.r - rect.l,
				rect.b - rect.t
			);
		}
	}

	addDecor(position: Point, type: string) {
		let decor = this.getDecor(position);
		if (decor) {
			this.removeDecor(position);
		}
		this.decor.push(
			new Decor({
				position,
				type,
				grid: this
			})
		);
	}
	getDecor(position: Point) {
		return this.decor.find(d => {
			return d.position.is(position);
		});
	}
	removeDecor(position: Point) {
		this.decor = this.decor.filter(d => {
			return !d.position.is(position);
		});
	}

	generate(seed: number) {
		let noise = new Noise(seed);
		this.blocks.forEach((col, x) => {
			col.forEach((block, y) => {
				let value = noise.simplex2(x / 50, y / 50);
				// console.log(x, y, value);
				value += y / col.length * 2 - 1;
				if (value > 0) {
					block.type = "1";
				}
			});
		});
	}
	height: number;
	width: number;

	fromTestStrings(strings: Array<string>): Grid {
		let testdata = strings.map(a => a.split(""));
		this.blocks = testdata[0].map(function(col, x) {
			return testdata.map(function(row, y) {
				return new Block({
					position: new Point({ x: x, y: y }),
					type: row[x],
					grid: this
				});
			});
		});
		return this;
	}

	constructor(size: { w: number, h: number } = { w: 10, h: 10 }) {
		super();
		this.tileCache = {};
		this.height = size.h;
		this.width = size.w;
		this.decor = [];
		this.z = -10;
		//make empty grid
		this.makeEmptyGrid(size);
		//new Block({ position: { x, y }, type: "0" })
	}
	makeEmptyGrid(size) {
		this.blocks = Array(size.w)
			.fill(0)
			.map((i, x) =>
				Array(size.h)
					.fill(0)
					.map(
						(j, y) =>
							new Block({
								position: new Point({ x, y }),
								type: "0",
								grid: this
							})
					)
			);
	}

	getBlock(pos: { x: number, y: number }): Block | void {
		if (this.blocks[pos.x]) {
			return this.blocks[pos.x][pos.y];
		}
	}
	getBlockAtPoint(point: { x: number, y: number }): Block | void {
		return this.blockAtPosition(point);
	}

	getBlocksFlattened = (): Array<Block> => {
		return this.blocks.reduce((a: Array<Block>, b: Array<Block>) => {
			// return a.splice(0, 0, ...b);
			return [].concat(a, b);
		}, []);
	};
	destroyBlockAtPosition(pos: { x: number, y: number }) {
		let x = Math.floor(pos.x / config.grid.width);
		let y = Math.floor(pos.y / config.grid.width);
		if (this.blocks[x]) {
			if (this.blocks[x][y]) {
				// this.grid[x][y] = "0";
				this.blocks[x][y].type = "0";
			}
		}
	}

	isPositionBlocked(pos: { x: number, y: number }) {
		let block = this.getBlockAtPoint(pos);
		return block && block.type != "0";
	}

	getBlocksInRect(rect: Rect): Array<Block> {
		let firstCol = Math.ceil(rect.l / config.grid.width);
		let lastCol = Math.floor(rect.r / config.grid.width);
		let firstRow = Math.ceil(rect.t / config.grid.width);
		let lastRow = Math.floor(rect.b / config.grid.width);

		if (lastCol < firstCol || lastRow < firstRow) {
			//return early
			return [];
		}
		return this.getBlockRect(firstCol, lastCol, firstRow, lastRow);
		// return out.filter(b => b !== undefined);
	}
	getBlockRect(
		firstCol: number,
		lastCol: number,
		firstRow: number,
		lastRow: number
	) {
		let out = [];
		for (let x = firstCol; x < lastCol; x++) {
			for (let y = firstRow; y < lastRow; y++) {
				let b = this.getBlock({ x, y });
				if (b !== undefined) {
					out.push(b);
				}
			}
		}
		return out;
	}

	getBlocksOverlappingRect(rect: Rect): Array<Block> {
		let firstCol = Math.floor(rect.l / config.grid.width);
		let lastCol = Math.ceil(rect.r / config.grid.width);
		let firstRow = Math.floor(rect.t / config.grid.width);
		let lastRow = Math.ceil(rect.b / config.grid.width);
		return this.getBlockRect(firstCol, lastCol, firstRow, lastRow);
	}

	blockAtPosition(pos: { x: number, y: number }): Block | void {
		let x = Math.floor(pos.x / config.grid.width);
		let y = Math.floor(pos.y / config.grid.width);
		//because y goes positive downwards, if an object is flat on the top
		//of a tile it will register as th e dtile below
		// console.log(pos.y, config.graid.waidth);
		if (pos.y % config.grid.width == 0) {
			y -= 1;
		}

		if (this.blocks[x]) {
			// return { block: this.grid[x][y], l: x * blocksize, t: y * blocksize };
			if (this.blocks[x][y]) {
				let block = this.blocks[x][y];
				return block;
			}
		}
		//TODO: remove this dummy
		return new Block({
			position: new Point({ x, y }),
			type: "1",
			grid: this
		});
	}

	// tileRenderer: TileRenderer;
	init(engine: Engine) {
		super.init(engine);
		engine.grid = this;
		// this.tileRenderer = new TileRenderer({ engine });
	}
	update = (engine: Engine) => {
		let screenRect = this.engine.ctx.screenRect();
		// console.log("wut");
		this.renderTiles(screenRect);
		this.renderDecor();
		// this.renderDebugBlockPixelLine();
	};
	tileCache: {};
	renderDebugBlockPixelLine() {
		let line = new Line({
			a: new Point({ x: 10.5, y: 10.5 }),
			b: this.engine.mouse.point.multiply(1 / config.grid.width)
		});
		this.highlightLine(line);
	}

	highlightLine(line) {
		let pixels = line.pixels();
		pixels.forEach(p => {
			this.highlightBlock(this.getBlock(p));
		});
		this.engine.ctx.drawLine(
			line.a.multiply(config.grid.width),
			line.b.multiply(config.grid.width),
			"red",
			2
		);
	}

	renderDecor() {
		this.engine.ctx.context.save();
		this.decor.forEach((decor: Decor) => {
			let type = decor.getType();
			this.engine.ctx.context.fillStyle = type.pattern;

			if (type.mode) {
				this.engine.ctx.context.globalCompositeOperation = type.mode;
			} else {
				this.engine.ctx.context.globalCompositeOperation =
					"source-over";
			}
			// this.engine.ctx.context.strokeStyle = "grey";
			//"#ff0000";
			this.engine.ctx.fillRectRelative(
				decor.position.x * config.grid.width,
				decor.position.y * config.grid.width,
				config.grid.width * type.width,
				config.grid.width * type.height
			);
			// this.engine.ctx.strokeRect(
			// 	decor.position.x * config.grid.width,
			// 	decor.position.y * config.grid.width,
			// 	config.grid.width * type.width,
			// 	config.grid.width * type.height
			// );
		});
		this.engine.ctx.context.restore();
	}
	renderTiles(screenRect) {
		//screenRect

		let newTileCache = {};
		this.tilesInRect(screenRect).forEach(tile => {
			// let tile = { x: 5, y: 6 };
			let tileImage = this.fetchTile(tile);
			this.engine.ctx.drawSprite(
				tileImage,
				new Point({
					x: tile.x * tileImage.width,
					y: tile.y * tileImage.height
				}),
				{ w: tileImage.width, h: tileImage.height },
				0,
				{ x: 0, y: 0 }
			);
			newTileCache[this.tileKey(tile)] = tileImage;
		});
		this.tileCache = newTileCache;
	}

	fetchTile(tile: { x: number, y: number }) {
		// if (!this.tileCache[this.tileKey(tile)]) {
		// 	this.tileCache[this.tileKey(tile)] = this.renderTile(tile);
		// }
		return this.tileCache[this.tileKey(tile)] || this.renderTile(tile);
	}
	tileKey(tile: { x: number, y: number }): string {
		return tile.x.toString() + "_" + tile.y.toString();
	}
	bustCache(block: Block) {
		this.tileCache[this.tileKey(this.tileForBlock(block))] = null;
	}
	tileForBlock(block: Block): { x: number, y: number } {
		return {
			x: Math.floor(block.position.x / config.grid.tile.width),
			y: Math.floor(block.position.y / config.grid.tile.height)
		};
	}

	tilesInRect(rect: Rect): Array<{ x: number, y: number }> {
		let w = config.grid.width * config.grid.tile.width;
		let h = config.grid.width * config.grid.tile.height;

		let firstCol = Math.floor(rect.l / w);
		let lastCol = Math.floor(rect.r / w);
		let firstRow = Math.floor(rect.t / h);
		let lastRow = Math.floor(rect.b / h);

		let out = [];
		for (let x = firstCol; x <= lastCol; x++) {
			for (let y = firstRow; y <= lastRow; y++) {
				out.push({ x, y });
			}
		}
		return out;
	}
	renderTile(tile: { x: number, y: number }): HTMLCanvasElement {
		let canvas: HTMLCanvasElement = document.createElement("canvas");
		canvas.width = config.grid.tile.width * config.grid.width;
		canvas.height = config.grid.tile.height * config.grid.width;

		// FLOWHACK
		let ctx: CanvasRenderingContext2D = canvas.getContext("2d");
		let offset = new Point({
			x: -canvas.width * tile.x,
			y: -canvas.height * tile.y
		});
		ctx.translate(offset.x, offset.y);
		let imageData = ctx.getImageData(
			0,
			0,
			ctx.canvas.width,
			ctx.canvas.height
		);

		for (let x = 0; x < config.grid.tile.width; x++) {
			for (let y = 0; y < config.grid.tile.height; y++) {
				let block = this.getBlock({
					x: tile.x * config.grid.tile.width + x,
					y: tile.y * config.grid.tile.height + y
				});
				if (block) {
					if (block.backgroundType !== "0") {
						let type = block.getBackgroundType();
						ctx.filter = "brightness(50%)";
						this.drawTile(ctx, type, x, y, offset, imageData);
					}
					if (!block.isEmpty()) {
						let type = block.getType();
						ctx.filter = "none";
						this.drawTile(ctx, type, x, y, offset, imageData);
					}
				}
				// if (block && block.backgroundType !== "0") {
				// 	let im = block.getBackgroundType().image;
				// 	ctx.filter = "brightness(50%)";
				// 	this.drawTile(ctx, im, x, y);
				// }
				// if (block && !block.isEmpty()) {
				// 	let im = block.getType().image;
				// 	// let imageParams = [im, 0, 0, im.width, im.height];
				// 	ctx.filter = "brightness(100%)";
				// 	this.drawTile(ctx, im, x, y);
				// }
			}
		}
		ctx.putImageData(imageData, 0, 0);
		ctx.strokeStyle = "rgba(0,0,0,0.1)";
		ctx.lineWidth = 1;
		ctx.strokeRect(-offset.x, -offset.y, canvas.width, canvas.height);
		//watermark
		ctx.font = "20px Verdana";
		// Fill with gradient
		ctx.fillStyle = "rgba(0,0,0,0.1)";
		ctx.fillText(tile.x + "," + tile.y, 8 - offset.x, 25 - offset.y);
		return canvas;
	}

	drawTile(ctx: CanvasRenderingContext2D, type, x, y, offset, imageData) {
		// let pattern = ctx.createPattern(im, "repeat");
		// let image

		// ctx.fillStyle = type.pattern;
		// ctx.fillRect(
		// 	x * config.grid.width - offset.x,
		// 	y * config.grid.width - offset.y,
		// 	config.grid.width,
		// 	config.grid.width
		// );

		// let imageData = ctx.getImageData(
		// 	0,
		// 	0,
		// 	ctx.canvas.width,
		// 	ctx.canvas.height
		// );
		let srcData = type.imageData.data;
		let destData = imageData.data;
		for (let dx = 0; dx < config.grid.width; dx++) {
			for (let dy = 0; dy < config.grid.width; dy++) {
				//FOR EACH PIXEL
				let dest = {
					x: dx + x * config.grid.width,
					y: dy + y * config.grid.width
				};
				let src = {
					x: dest.x % type.image.width,
					y: dest.y % type.image.height
				};
				let srcloc = (src.y * type.image.width + src.x) * 4;
				let destloc = (dest.y * ctx.canvas.width + dest.x) * 4;
				destData[destloc] = srcData[srcloc];
				destData[destloc + 1] = srcData[srcloc + 1];
				destData[destloc + 2] = srcData[srcloc + 2];
				destData[destloc + 3] = srcData[srcloc + 3];
			}
		}
		// ctx.putImageData(imageData, 0, 0);
	}

	save(): string {
		// return JSON.stringify({ blocks: this.blocks }, (name, val) => {
		// 	if (name !== "grid") return val;
		// });
		let enemies = [];
		if (this.engine) {
			enemies = this.engine.objectsTagged("enemy");
		}
		// console.log(enemies);
		return JSON.stringify({
			enemies: enemies.map(e => {
				return { t: e.type.id, p: e.position };
			}),
			decor: this.decor.map(d => {
				return { t: d.type, p: d.position };
			}),
			blocks: this.blocks.map(col => {
				return col.map(block => {
					return { t: block.type, b: block.backgroundType };
				});
			})
		});
	}
	load(str: string) {
		//wipe enemies
		if (this.engine) {
			this.engine.objectsTagged("enemy").forEach(e => e.destroy());
		}

		this.tileCache = {};
		let data = JSON.parse(str);
		if (data.decor) {
			this.decor = data.decor.map(decor => {
				return new Decor({
					position: new Point(decor.p),
					type: decor.t,
					grid: this
				});
			});
		}
		this.blocks = data.blocks.map((d, x) => {
			return d.map((block, y) => {
				return new Block({
					position: new Point({ x, y }),
					type: block.t,
					backgroundType: block.b,
					grid: this
				});
			});
		});
		if (data.enemies) {
			data.enemies.forEach(e => {
				let type = EnemyTypesMap[e.t];
				if (!type) {
					throw new Error("could not look up enemy", e.t);
				}
				this.engine.register(
					new Enemy({
						position: new Point(e.p),
						type: type
					})
				);
			});
		}
	}
	addRowAbove() {
		this.height++;
		this.blocks.forEach(col => {
			col.unshift(new Block(col[0]));
		});
		this.rebuildBlocks();
	}
	addRowBelow() {
		this.height++;
		this.blocks.forEach(col => {
			col.push(new Block(col[col.length - 1]));
		});
		this.rebuildBlocks();
	}
	addColLeft() {
		this.width++;
		this.blocks.unshift(this.blocks[0].map(b => new Block(b)));
		this.rebuildBlocks();
	}
	addColRight() {
		this.width++;
		this.blocks.push(
			this.blocks[this.blocks.length - 1].map(b => new Block(b))
		);
		this.rebuildBlocks();
	}
	rebuildBlocks() {
		this.tileCache = {};
		this.blocks.forEach((col, x) =>
			col.forEach((block, y) => (block.position = new Point({ x, y })))
		);
	}
}
