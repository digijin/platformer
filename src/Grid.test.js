//@flow

import Grid from "Grid";
import config from "config";
import Rect from "Utility/Rect";

describe("getBlocksInRect", () => {
	describe("should return", () => {
		let grid, rect, blocks;

		beforeEach(() => {
			config.grid.width = 10;

			grid = new Grid({ size: { w: 10, h: 10 } });
			rect = new Rect({ t: 15, r: 45, b: 45, l: 15 });
			blocks = grid.getBlocksInRect(rect);
		});
		it("should return array", () => {
			expect(Array.isArray(blocks)).toBe(true);
		});
		it("should have right number of elements", () => {
			expect(blocks.length).toBe(4);
		});
		it("sohuld have correct elements", () => {
			expect(blocks[0].position.x).toBe(2);
			expect(blocks[0].position.y).toBe(2);

			expect(blocks[3].position.x).toBe(3);
			expect(blocks[3].position.y).toBe(3);
		});
	});
});
describe("Grid.test.js jest only", () => {
	describe("save", () => {
		it("should return a strugn from save", () => {
			const grid = new Grid({ size: { w: 10, h: 10 } });
			expect(typeof grid.save()).toBe("string");
		});
		it("save snapshot", () => {
			const grid = new Grid({ size: { w: 2, h: 2 } });
			// FLOWHACK
			expect(grid.save()).toMatchSnapshot();
		});
	});
});
