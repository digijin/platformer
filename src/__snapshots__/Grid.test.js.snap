exports[`grid jest save save snapshot 1`] = `"{\"enemies\":[],\"decor\":[],\"blocks\":[[{\"t\":\"0\",\"b\":\"0\"},{\"t\":\"0\",\"b\":\"0\"}],[{\"t\":\"0\",\"b\":\"0\"},{\"t\":\"0\",\"b\":\"0\"}]]}"`;
